File: cpp-pthread/exer07-exec-service-main.cpp
/*
EXECUTOR SERVICE & THREAD POOL IMPLEMENTATION
*/


#include <iostream>
#include <unistd.h>
#include "exer07-exec-service-itask.hpp"
#include "exer07-exec-service-v1a.hpp"
#include "exer07-exec-service-v1b.hpp"
#include "exer07-exec-service-v2a.hpp"
#include "exer07-exec-service-v2b.hpp"



class MyTask : public ITask {
public:
    char id;

public:
    void run() override {
        std::cout << "Task " << id << " is starting" << std::endl;
        sleep(3);
        std::cout << "Task " << id << " is completed" << std::endl;
    }
};



int main() {
    constexpr int NUM_THREADS = 2;
    constexpr int NUM_TASKS = 5;


    MyExecServiceV2B execService;
    execService.init(NUM_THREADS);


    std::vector<MyTask> lstTask(NUM_TASKS);

    for (int i = 0; i < NUM_TASKS; ++i)
        lstTask[i].id = 'A' + i;


    for (auto&& task : lstTask)
        execService.submit(&task);

    std::cout << "All tasks are submitted" << std::endl;


    execService.waitTaskDone();
    std::cout << "All tasks are completed" << std::endl;


    execService.shutdown();


    return 0;
}

File: csharp/exercise/exer07-exec-service-main.cs
/*
 * EXECUTOR SERVICE & THREAD POOL IMPLEMENTATION
 */
using System;
using System.Collections.Generic;
using System.Threading;



namespace Exer07
{
    class MainApp : IRunnable
    {
        public void run()
        {
            const int NUM_THREADS = 2;
            const int NUM_TASKS = 5;


            var execService = new MyExecServiceV0A(NUM_THREADS);


            var lstTask = new List<MyTask>();

            for (int i = 0; i < NUM_TASKS; ++i)
                lstTask.Add(new MyTask((char)('A' + i)));


            lstTask.ForEach(task => execService.submit(task));
            Console.WriteLine("All tasks are submitted");


            execService.waitTaskDone();
            Console.WriteLine("All tasks are completed");


            execService.shutdown();
        }
    }



    class MyTask : IRunnable
    {
        public char id;

        public MyTask(char id) {
            this.id = id;
        }

        public void run() {
            Console.WriteLine($"Task {id} is starting");
            Thread.Sleep(3000);
            Console.WriteLine($"Task {id} is completed");
        }
    }
}
